### PR TITLE
Qlib Highfreq Support & Highfreq DataHanlder/Operator/Processor Examples

### PR DESCRIPTION
**Pull request by bxdd**
## Description

Qlib Highfreq Support & Highfreq DataHanlder/Operator/Processor Examples
## How Has This Been Tested?
- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [x] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: examples/high_freq/workflow.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                    "test": (
+                        test_start_time,
+                        end_time,
+                    ),
+                },
+            },
+        },
+    }
+    ##=============load the calendar for cache=============
+    # unnecessary, but may accelerate
```
Comment (bxdd) [inline examples/high_freq/workflow.py:111]: Is there a better way to load the cache in advance? @you-n-g

File: qlib/data/data.py [hunk 2]
```diff
@@ -122,6 +122,16 @@ def _get_calendar(self, freq, future):
             H["c"][flag] = _calendar, _calendar_index
         return _calendar, _calendar_index
 
+    def get_calendar_day(self, freq="day", future=False):
```
Comment (bxdd) [inline qlib/data/data.py:125]: `get_calendar_day` just for high-freq, if there are other ways to implement it

File: examples/high_freq/highfreq_processor.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            slice0 = part_values > 3.0
+            slice1 = part_values > 3.5
+            slice2 = part_values < -3.0
+            slice3 = part_values < -3.5
+
+            part_values[slice0] = 3.0 + (part_values[slice0] - 3.0) / (self.feature_vmax[name] - 3) * 0.5
+            part_values[slice1] = 3.5
+            part_values[slice2] = -3.0 - (part_values[slice2] + 3.0) / (self.feature_vmin[name] + 3) * 0.5
+            part_values[slice3] = -3.5
+        # print("start_call_feature_reshape")
```
Comment (you-n-g) [inline examples/high_freq/highfreq_processor.py:58]: remove useless line or  replace it with log

File: examples/high_freq/highfreq_processor.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        idx = df_features.index.droplevel("datetime").drop_duplicates()
+        idx.set_names(["instrument", "datetime"], inplace=True)
+        feat = df_values[:, [0, 1, 2, 3, 4, 10]].reshape(-1, 6 * 240)
```
Comment (you-n-g) [inline examples/high_freq/highfreq_processor.py:61]: Add some comments about  this weird behaviour

File: examples/high_freq/workflow.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+from qlib.data.dataset.handler import DataHandlerLP
+from qlib.data.ops import Operators
+from qlib.data.data import Cal
+
+from highfreq_ops import DayFirst, DayLast, FFillNan, Date, Select, IsNull
+
+if __name__ == "__main__":
+
+    # use default data
+    provider_uri = "/nfs_data/qlib_data/yahoo_high_qlib"  # target_dir
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply examples/high_freq/workflow.py]: replace it with qlib offline dataset

File: examples/high_freq/workflow.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    qlib.init(
+        provider_uri=provider_uri,
+        custom_ops=[DayFirst, DayLast, FFillNan, Date, Select, IsNull],
+        redis_port=-1,
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:42]: dataset_cache=None instead of  `redis_port=-1`

File: examples/high_freq/workflow.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+from qlib.data.data import Cal
+
+from highfreq_ops import DayFirst, DayLast, FFillNan, Date, Select, IsNull
+
+if __name__ == "__main__":
+
+    # use default data
+    provider_uri = "/nfs_data/qlib_data/yahoo_high_qlib"  # target_dir
+    qlib.init(
+        provider_uri=provider_uri,
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:40]: Maybe we can add  "HIGH_FREQ_CONFIG"  in config.py?

File: examples/high_freq/workflow.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        redis_port=-1,
+        region=REG_CN,
+        auto_mount=False,
+    )
+
+    MARKET = "all"
+    BENCHMARK = "SH000300"
+    DROP_LOAD_DATASET = False  # flag wether to test [drop and load dataset]
+
+    # start_time = "2019-01-01 00:00:00"
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:51]: Remove useless lines

File: examples/high_freq/workflow.py (hunk 1 above)
Comment (you-n-g) [reply examples/high_freq/workflow.py:119]: I don't have better ideas so far....
We can add some comments to explain that this acceleration doesn't work on windows and mac .

File: examples/high_freq/workflow.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    print(backtest_train, backtest_test)
+
+    del xtrain, xtest
+    del backtest_train, backtest_test
+
+    ## example to show how to save the dataset and reload it, and how to use different data
+    if DROP_LOAD_DATASET:
+
+        ##=============dump dataset=============
+        dataset.to_pickle(path="dataset.pkl")
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:140]: This example is hard to use

File: examples/high_freq/workflow.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        dataset_backtest.to_pickle(path="dataset_backtest.pkl")
+
+        del dataset, dataset_backtest
+        ##=============reload dataset=============
+        file_dataset = open("dataset.pkl", "rb")
+        dataset = pickle.load(file_dataset)
+        file_dataset.close()
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:147]: with statement is preferred

File: examples/workflow_by_code.py [hunk 11]
```diff
@@ -30,7 +30,7 @@
 
         GetData().qlib_data(target_dir=provider_uri, region=REG_CN)
 
-    qlib.init(provider_uri=provider_uri, region=REG_CN)
+    qlib.init(provider_uri=provider_uri, region=REG_CN, redis_port=233)
```
Comment (you-n-g) [inline examples/workflow_by_code.py:33]: 2333333

File: qlib/data/data.py (hunk 2 above)
Comment (you-n-g) [reply qlib/data/data.py:125]: We don't need to put this in `qlib/data/data.py`

File: qlib/data/dataset/__init__.py [hunk 12]
```diff
@@ -87,6 +87,17 @@ def __init__(self, handler: Union[dict, DataHandler], segments: list):
         """
         super().__init__(handler, segments)
 
+    def init(self, **kwargs):
+
+        logger = get_module_logger("DatasetH")
+        handler_init_kwargs = {}
+        for arg_key, arg_value in kwargs.items():
```
Comment (you-n-g) [inline qlib/data/dataset/__init__.py:94]: it is a little weird

Comment (you-n-g) [reply qlib/data/dataset/__init__.py:94]: Let's discuss it later

Comment (you-n-g) [reply qlib/data/dataset/__init__.py:94]: arguments  like `*kwargs`  is not supported in such implementation

File: qlib/data/dataset/handler.py [hunk 13]
```diff
@@ -416,7 +428,13 @@ def init(self, init_type: str = IT_FIT_SEQ, enable_cache: bool = False):
         # TODO: Be able to cache handler data. Save the memory for data processing
 
     def _get_df_by_key(self, data_key: str = DK_I) -> pd.DataFrame:
-        df = getattr(self, {self.DK_R: "_data", self.DK_I: "_infer", self.DK_L: "_learn"}[data_key])
+        try:
+            df = getattr(self, {self.DK_R: "_data", self.DK_I: "_infer", self.DK_L: "_learn"}[data_key])
+        except AttributeError:
+            print("please set drop_raw = False if you want to use raw data")
+            raise
+        except:
```
Comment (you-n-g) [inline qlib/data/dataset/handler.py:436]: What is it for?

File: qlib/data/dataset/handler.py [hunk 14]
```diff
@@ -416,7 +428,13 @@ def init(self, init_type: str = IT_FIT_SEQ, enable_cache: bool = False):
         # TODO: Be able to cache handler data. Save the memory for data processing
 
     def _get_df_by_key(self, data_key: str = DK_I) -> pd.DataFrame:
-        df = getattr(self, {self.DK_R: "_data", self.DK_I: "_infer", self.DK_L: "_learn"}[data_key])
+        try:
+            df = getattr(self, {self.DK_R: "_data", self.DK_I: "_infer", self.DK_L: "_learn"}[data_key])
+        except AttributeError:
+            print("please set drop_raw = False if you want to use raw data")
```
Comment (you-n-g) [inline qlib/data/dataset/handler.py:434]: You should raise error when only `_data`   is raised

File: qlib/data/ops.py [hunk 15]
```diff
@@ -17,11 +17,13 @@
 try:
     from ._libs.rolling import rolling_slope, rolling_rsquare, rolling_resi
     from ._libs.expanding import expanding_slope, expanding_rsquare, expanding_resi
-except ImportError as err:
+except ImportError:
     print(
         "#### Do not import qlib package in the repository directory in case of importing qlib from . without compiling #####"
     )
     raise
+except:
```
Comment (you-n-g) [inline qlib/data/ops.py:25]: What is this except for?

File: examples/high_freq/workflow.py (hunk 6 above)
Comment (you-n-g) [reply examples/high_freq/workflow.py:42]: Default in most case
- Enable expression cache , disable dataset cache.

Comment (you-n-g) [reply examples/high_freq/workflow.py:42]: disable expression cache in this case

File: examples/high_freq/workflow.py (hunk 7 above)
Comment (you-n-g) [reply examples/high_freq/workflow.py:40]: SPEC_CONF
HIGH_FREQ_CONFIG

``qlib.init(**HIGH_FREQ_CONFIG,  **SPEC_CONF)`

Comment (you-n-g) [reply examples/high_freq/workflow.py:40]: provider_uri in HIGH_FREQ_CONFIG

File: examples/high_freq/workflow.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+        file_dataset_backtest = open("dataset_backtest.pkl", "rb")
+        dataset_backtest = pickle.load(file_dataset_backtest)
+
+        file_dataset_backtest.close()
```
Comment (you-n-g) [inline examples/high_freq/workflow.py:152]: same as above

File: qlib/data/data.py [hunk 17]
```diff
@@ -669,7 +679,10 @@ def expression(self, instrument, field, start_time=None, end_time=None, freq="da
         # 1) The stock data is currently float. If there is other types of data, this part needs to be re-implemented.
         # 2) The the precision should be configurable
         try:
-            series = series.astype(np.float32)
+            if series.dtype == np.float64:
+                series = series.astype(np.float32)
+            elif series.dtype == np.bool:
+                series = series.astype(np.int8)
```
Comment (you-n-g) [inline qlib/data/data.py:685]: else?

Comment (you-n-g) [reply qlib/data/data.py:685]: Add comments

Comment (you-n-g) [reply qlib/data/data.py:685]: if special_case:
   XXXX
else:
   default behavior

File: qlib/data/dataset/__init__.py (hunk 12 above)
Comment (you-n-g) [reply qlib/data/dataset/__init__.py:94]: DatasetH.init(dh_kwargs)

self.handler.init(**dh_kwargs)

File: examples/highfreq/highfreq_handler.py [hunk 18]
```diff
@@ -56,88 +56,44 @@ def get_feature_config(self):
...
-                    template_paused.format("$low"),
-                ),
-                template_fillnan.format(template_paused.format("$close")),
-            )
-        ]
-        fields += ["{0}/Ref(DayLast({0}), 240)".format(template_fillnan.format(template_paused.format("$close")))]
-        fields += [
-            "{0}/Ref(DayLast({1}), 240)".format(
+
+        def get_04_price_feature(price_field):
```
Comment (you-n-g) [inline examples/highfreq/highfreq_handler.py:63]: `get_04_price_feature` &  `get_59_price_feature` are really weird names.

File: examples/highfreq/highfreq_ops.py [hunk 19]
```diff
@@ -3,51 +3,61 @@
...
-        _calendar = Cal.get_calendar_day(freq=freq)[0]
+        _calendar = get_calendar_day(freq=freq)
         series = self.feature.load(instrument, start_index, end_index, freq)
         return series.groupby(_calendar[series.index]).transform("last")
 
 
 class FFillNan(ElemOperator):
     def __init__(self, feature):
-        super(FFillNan, self).__init__(feature, "fill_nan")
+        super(FFillNan, self).__init__(feature)
```
Comment (you-n-g) [inline examples/highfreq/highfreq_ops.py:32]: These `init` function are no longer required.

File: examples/highfreq/highfreq_processor.py [hunk 20]
```diff
@@ -41,23 +41,27 @@ def __call__(self, df_features):
...
-            part_values = df_values[:, name_val]
             if name == "volume":
-                part_values[:] = np.log1p(part_values)
-            part_values -= self.feature_med[name]
-            part_values /= self.feature_std[name]
-            slice0 = part_values > 3.0
-            slice1 = part_values > 3.5
-            slice2 = part_values < -3.0
-            slice3 = part_values < -3.5
+                df_values[:, name_val] = np.log1p(df_values[:, name_val])
```
Comment (you-n-g) [inline examples/highfreq/highfreq_processor.py:45]: What's the difference here ?

File: examples/highfreq/workflow.py [hunk 21]
```diff
@@ -115,64 +94,77 @@
...
-    xtrain, xtest = dataset.prepare(["train", "test"])
-    print(xtrain, xtest)
+    def _init_qlib(self):
+        """initialize qlib"""
+        # use yahoo_cn_1min data
+        QLIB_INIT_CONFIG = {**HIGH_FREQ_CONFIG, **self.SPEC_CONF}
+        provider_uri = QLIB_INIT_CONFIG.get("provider_uri")
+        if not exists_qlib_data(provider_uri):
+            print(f"Qlib data is not found in {provider_uri}")
+            sys.path.append(str(Path(__file__).resolve().parent.parent.parent.joinpath("scripts")))
```
Comment (you-n-g) [inline examples/highfreq/workflow.py:105]: `from qlib.tests.data import GetData`

File: examples/highfreq/highfreq_handler.py (hunk 18 above)
Comment (you-n-g) [reply examples/highfreq/highfreq_handler.py:63]: get_normalized_price_fea

File: examples/workflow_by_code.py [hunk 22]
```diff
@@ -17,20 +17,17 @@
...
     provider_uri = "~/.qlib/qlib_data/cn_data"  # target_dir
     if not exists_qlib_data(provider_uri):
         print(f"Qlib data is not found in {provider_uri}")
-        sys.path.append(str(Path(__file__).resolve().parent.parent.joinpath("scripts")))
-        from get_data import GetData
-
         GetData().qlib_data(target_dir=provider_uri, region=REG_CN)
 
-    qlib.init(provider_uri=provider_uri, region=REG_CN)
+    qlib.init(provider_uri=provider_uri, region=REG_CN, redis_port=-1)
```
Comment (you-n-g) [inline examples/workflow_by_code.py:30]: Hello

**Pull request merged**